operators/katib-controller/src/charm.py

```diff
@@ -0,0 +1,309 @@
+#!/usr/bin/env python3
+
+import logging
+from pathlib import Path
+from subprocess import check_call
+
+import yaml
+from ops.charm import CharmBase
+from ops.main import main
+from ops.model import ActiveStatus, MaintenanceStatus
+from ops.framework import StoredState
+
+from oci_image import OCIImageResource, OCIImageResourceError
+
+logger = logging.getLogger(__name__)
+
+
+class Operator(CharmBase):
+    """Deploys the katib-controller service."""
+
+    _stored = StoredState()
+
+    def __init__(self, framework):
+        super().__init__(framework)
+
+        if not self.model.unit.is_leader():
+            logger.info("Not a leader, skipping any work")
+            self.model.unit.status = ActiveStatus()
+            return
+
+        self._stored.set_default(**self.gen_certs())
+        self.image = OCIImageResource(self, "oci-image")
+        self.framework.observe(self.on.install, self.set_pod_spec)
+        self.framework.observe(self.on.upgrade_charm, self.set_pod_spec)
+
+    def set_pod_spec(self, event):
+        self.model.unit.status = MaintenanceStatus("Setting pod spec")
+
+        try:
+            image_details = self.image.fetch()
+        except OCIImageResourceError as e:
+            self.model.unit.status = e.status
+            return
+
+        validating, mutating = yaml.safe_load_all(Path("src/webhooks.yaml").read_text())
+
+        self.model.pod.set_spec(
```

Comment:
I think it would be great to create script to automatically generate this spec from the appropriate YAML manifests.

Reply:
I've opened this issue to talk about supporting it in the operator framework directly:

https://github.com/canonical/operator/issues/484